File: python/src/linked_list/reverse_linked_list.py
from list import List


def reverse_list(head: List) -> List:
    prev = None
    cursor = head
    while cursor:
        next = cursor.next
        cursor.next = prev
        prev = cursor
        cursor = next
    return prev


def reverse_list_recursive(head: List) -> List:
    if not head or not head.next:
        return head

    new_head = reverse_list_recursive(head.next)
    head.next.next = head
    head.next = None
    return new_head

File: typescript/src/linked-list/reverse-linked-list.ts
import { List, ListNode } from "./list";

function reverseList(head: List): List {
  let prev = null;
  let cursor = head;
  while (cursor) {
    const next = cursor.next;
    cursor.next = prev;
    prev = cursor;
    cursor = next;
  }
  return prev;
}

function reverseListRecursive(head: List): List {
  if (!head?.next) {
    return head;
  }
  const newHead = reverseListRecursive(head.next);
  head.next.next = head;
  head.next = null;
  return newHead;
}

if (import.meta.vitest) {
  const { test, expect } = await import("vitest");

  test("reverseList", () => {
    const list = ListNode.from_values([1, 2, 3, 4, 5]);
    const reversed = reverseList(list);
    expect(Array.from(reversed!.values())).toEqual([5, 4, 3, 2, 1]);
  });

  test("reverseListRecursive", () => {
    const list = ListNode.from_values([1, 2, 3, 4, 5]);
    const reversed = reverseListRecursive(list);
    expect(Array.from(reversed!.values())).toEqual([5, 4, 3, 2, 1]);
  });
}
